File: 2020/day15/part1/part2.py
import time

input = [9, 19, 1, 6, 0, 5, 4]

#input = [0, 3, 6]

#input = [3, 1, 2]

N = 30000000
#N = 100

t_start = time.time()

turn_number_spoken = {}

for i, number in enumerate(input[:-1]):
    turn_number_spoken[number] = i + 1

last_spoken = input[-1]
for turn in range(len(input) + 1, N + 1):
    last_turn = turn_number_spoken.get(last_spoken)
    if last_turn:
        n = turn - 1 - last_turn
    else:
        n = 0
    turn_number_spoken[last_spoken] = turn - 1
    last_spoken = n
print(n)
print(time.time() - t_start)

File: 2020/day15/part1/part2.cpp
#include <iostream>
#include <vector>

const int N = 30000000;

int turn_number[N];

int main() {
    //std::vector<int> input = {0, 3, 6};
    std::vector<int> input = {9, 19, 1, 6, 0, 5, 4};

    for (size_t i = 0; i < input.size() - 1; i++) {
        turn_number[input[i]] = i + 1;
    }

    int last_spoken = input[input.size() - 1];

    int n;
    for (int turn = input.size() + 1; turn < N + 1; turn++) {
        int last_turn = turn_number[last_spoken];
        turn_number[last_spoken] = turn - 1;
        n = 0;
        if (last_turn) {
            n = turn - 1 - last_turn;
        }
        last_spoken = n;
    }

    std::cout << n << std::endl;
    
    return 0;
}
